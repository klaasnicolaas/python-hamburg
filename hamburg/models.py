"""Asynchronous Python client providing Urban Data information of Hamburg."""
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
from typing import Any


@dataclass
class DisabledParking:
    """Object representing a disabled parking."""

    spot_id: str
    street: str | None
    limitation: str | None
    number: int
    longitude: float
    latitude: float

    @classmethod
    def from_json(cls, data: dict[str, Any]) -> DisabledParking:
        """Return a DisabledParking object from a JSON dictionary.

        Args:
            data: The JSON data from the API.

        Returns:
            A DisabledParking object.
        """

        def strip_spaces(string: str) -> str | None:
            """Strip spaces from a string.

            Args:
                string: The string to strip.

            Returns:
                The string without spaces or None if the string is empty.
            """
            if string is None:
                return None
            return string.strip()

        attr = data["properties"]
        geo = data["geometry"]["coordinates"]
        return cls(
            spot_id=str(data.get("id")),
            street=strip_spaces(attr.get("nahe_adresse")),
            limitation=strip_spaces(attr.get("befristung")),
            number=attr.get("anzahl"),
            longitude=geo[0],
            latitude=geo[1],
        )


@dataclass
class ParkAndRide:
    """Object representing a park and ride spot."""

    spot_id: str
    name: str
    construction_year: int
    address: str
    public_transport_line: str
    park_type: str
    free_space: int
    capacity: int
    occupancy_percentage: float
    disabled_parking_spaces: int
    tickets: dict[str, int]
    longitude: float
    latitude: float
    updated_at: datetime

    @classmethod
    def from_json(cls, data: dict[str, Any]) -> ParkAndRide:
        """Return a ParkAndRide object from a JSON dictionary.

        Args:
            data: The JSON data from the API.

        Returns:
            A ParkAndRide object.
        """

        attr = data["properties"]
        geo = data["geometry"]["coordinates"]
        return cls(
            spot_id=str(data.get("id")),
            name=attr.get("name"),
            construction_year=attr.get("baujahr"),
            address=attr.get("adresse"),
            public_transport_line=attr.get("linie"),
            park_type=attr.get("art"),
            free_space=attr.get("stellplaetze_frei"),
            capacity=attr.get("stellplaetze_gesamt"),
            occupancy_percentage=round(attr.get("stellpl_frei_in_prozent"), 2),
            disabled_parking_spaces=attr.get("stellplaetze_behinderte_gesamt"),
            tickets={
                "day": attr.get("ticket_1_tag"),
                "month": attr.get("ticket_30_tage"),
                "year": attr.get("ticket_1_jahr"),
            },
            longitude=geo[0],
            latitude=geo[1],
            updated_at=datetime.strptime(
                attr.get("aktualitaet_belegungsdaten"), "%Y-%m-%d %H:%M:%S"
            ),
        )
