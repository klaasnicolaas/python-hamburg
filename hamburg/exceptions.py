"""Asynchronous Python client providing Urban Data information of Hamburg."""


class UDPHamburgError(Exception):
    """Generic Urban Data Platform Hamburg exception."""


class UDPHamburgConnectionError(UDPHamburgError):
    """Urban Data Platform Hamburg - connection error."""
