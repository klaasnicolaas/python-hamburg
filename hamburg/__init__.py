"""Asynchronous Python client providing Urban Data information of Hamburg."""

from .exceptions import UDPHamburgConnectionError, UDPHamburgError
from .hamburg import UDPHamburg
from .models import DisabledParking, ParkAndRide

__all__ = [
    "UDPHamburg",
    "UDPHamburgConnectionError",
    "UDPHamburgError",
    "DisabledParking",
    "ParkAndRide",
]
