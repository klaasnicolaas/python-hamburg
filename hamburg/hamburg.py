"""Asynchronous Python client providing Urban Data information of Hamburg."""
from __future__ import annotations

import asyncio
import socket
from dataclasses import dataclass
from importlib import metadata
from typing import Any

import aiohttp
import async_timeout
from aiohttp import hdrs
from yarl import URL

from .exceptions import UDPHamburgConnectionError, UDPHamburgError
from .models import DisabledParking, ParkAndRide


@dataclass
class UDPHamburg:
    """Main class for handling data fetchting from Urban Data Platform of Hamburg."""

    request_timeout: float = 10.0
    session: aiohttp.client.ClientSession | None = None

    _close_session: bool = False

    async def _request(
        self,
        uri: str,
        *,
        method: str = hdrs.METH_GET,
        params: dict[str, Any] | None = None,
    ) -> Any:
        version = metadata.version(__package__)
        url = URL.build(
            scheme="https", host="api.hamburg.de", path="/datasets/v1/"
        ).join(URL(uri))

        headers = {
            "Accept": "application/geo+json",
            "User-Agent": f"PythonUDPHamburg/{version}",
        }

        if self.session is None:
            self.session = aiohttp.ClientSession()
            self._close_session = True

        try:
            async with async_timeout.timeout(self.request_timeout):
                response = await self.session.request(
                    method,
                    url,
                    params=params,
                    headers=headers,
                    ssl=True,
                )
                response.raise_for_status()
        except asyncio.TimeoutError as exception:
            raise UDPHamburgConnectionError(
                "Timeout occurred while connecting to the Urban Data Platform API."
            ) from exception
        except (aiohttp.ClientError, socket.gaierror) as exception:
            raise UDPHamburgConnectionError(
                "Error occurred while communicating with Urban Data Platform API."
            ) from exception

        content_type = response.headers.get("Content-Type", "")
        if "application/geo+json" not in content_type:
            text = await response.text()
            raise UDPHamburgError(
                "Unexpected content type response from the Urban Data Platform API",
                {"Content-Type": content_type, "Response": text},
            )

        return await response.json()

    async def disabled_parkings(
        self, limit: int = 10, bulk: str = "false"
    ) -> list[DisabledParking]:
        """Get all disabled parking spaces.

        Args:
            limit: Number of items to return.
            bulk: Whether to return all items or the limit.

        Returns:
            A list of DisabledParking objects.
        """

        results: list[DisabledParking] = []
        locations = await self._request(
            "behindertenstellplaetze/collections/verkehr_behindertenparkpl/items",
            params={"limit": limit, "bulk": bulk},
        )

        for item in locations["features"]:
            results.append(DisabledParking.from_json(item))
        return results

    async def park_and_ride(
        self, limit: int = 10, bulk: str = "false"
    ) -> list[ParkAndRide]:
        """Get all park and ride spaces.

        Args:
            limit: Number of items to return.
            bulk: Whether to return all items or the limit.

        Returns:
            A list of ParkAndRide objects.
        """
        results: list[ParkAndRide] = []
        locations = await self._request(
            "p_und_r/collections/p_und_r/items",
            params={"limit": limit, "bulk": bulk},
        )
        print(locations)
        for item in locations["features"]:
            results.append(ParkAndRide.from_json(item))
        return results

    async def close(self) -> None:
        """Close open client session."""
        if self.session and self._close_session:
            await self.session.close()

    async def __aenter__(self) -> UDPHamburg:
        """Async enter.

        Returns:
            The Urban Data Platform object.
        """
        return self

    async def __aexit__(self, *_exc_info: str) -> None:
        """Async exit.

        Args:
            _exc_info: Exec type.
        """
        await self.close()
