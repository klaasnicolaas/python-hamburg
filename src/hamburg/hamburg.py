"""Asynchronous Python client providing Urban Data information of Hamburg."""
from __future__ import annotations

import asyncio
import socket
from dataclasses import dataclass
from importlib import metadata
from typing import Any, cast

import async_timeout
from aiohttp import ClientError, ClientSession
from aiohttp.hdrs import METH_GET
from yarl import URL

from .exceptions import UDPHamburgConnectionError, UDPHamburgError
from .models import DisabledParking, Garage, ParkAndRide


@dataclass
class UDPHamburg:
    """Main class for handling data fetchting from Urban Data Platform of Hamburg."""

    request_timeout: float = 10.0
    session: ClientSession | None = None

    _close_session: bool = False

    async def _request(
        self,
        uri: str,
        *,
        method: str = METH_GET,
        params: dict[str, Any] | None = None,
    ) -> Any:
        """Handle a request to the Urban Data Platform API of Hamburg.

        Args:
        ----
            uri: Request URI, without '/', for example, 'status'
            method: HTTP method to use, for example, 'GET'
            params: Extra options to improve or limit the response.

        Returns:
        -------
            A Python dictionary (text) with the response from
            the Urban Data Platform API.

        Raises:
        ------
            UDPHamburgConnectionError: Timeout occurred while
                connecting to the Urban Data Platform API.
            UDPHamburgError: If the data is not valid.
        """
        version = metadata.version(__package__)
        url = URL.build(
            scheme="https",
            host="api.hamburg.de",
            path="/datasets/v1/",
        ).join(URL(uri))

        headers = {
            "Accept": "application/geo+json",
            "User-Agent": f"PythonUDPHamburg/{version}",
        }

        if self.session is None:
            self.session = ClientSession()
            self._close_session = True

        try:
            async with async_timeout.timeout(self.request_timeout):
                response = await self.session.request(
                    method,
                    url,
                    params=params,
                    headers=headers,
                    ssl=True,
                )
                response.raise_for_status()
        except asyncio.TimeoutError as exception:
            msg = "Timeout occurred while connecting to the Urban Data Platform API."
            raise UDPHamburgConnectionError(
                msg,
            ) from exception
        except (ClientError, socket.gaierror) as exception:
            msg = "Error occurred while communicating with Urban Data Platform API."
            raise UDPHamburgConnectionError(
                msg,
            ) from exception

        content_type = response.headers.get("Content-Type", "")
        if "application/geo+json" not in content_type:
            text = await response.text()
            msg = "Unexpected content type response from the Urban Data Platform API"
            raise UDPHamburgError(
                msg,
                {"Content-Type": content_type, "Response": text},
            )

        return cast(dict[str, Any], await response.json())

    async def disabled_parkings(
        self,
        limit: int = 10,
        bulk: str = "false",
    ) -> list[DisabledParking]:
        """Get all disabled parking spaces.

        Args:
        ----
            limit: Number of items to return.
            bulk: Whether to return all items or the limit.

        Returns:
        -------
            A list of DisabledParking objects.
        """
        locations = await self._request(
            "behindertenstellplaetze/collections/verkehr_behindertenparkpl/items",
            params={"limit": limit, "bulk": bulk},
        )
        return [DisabledParking.from_dict(item) for item in locations["features"]]

    async def park_and_rides(
        self,
        limit: int = 10,
        bulk: str = "false",
    ) -> list[ParkAndRide]:
        """Get all park and ride spaces.

        Args:
        ----
            limit: Number of items to return.
            bulk: Whether to return all items or the limit.

        Returns:
        -------
            A list of ParkAndRide objects.
        """
        locations = await self._request(
            "p_und_r/collections/p_und_r/items",
            params={"limit": limit, "bulk": bulk},
        )
        return [ParkAndRide.from_dict(item) for item in locations["features"]]

    async def garages(
        self,
        limit: int = 10,
        bulk: str = "false",
        available: str | None = None,
    ) -> list[Garage]:
        """Get all garages.

        Args:
        ----
            limit: Number of items to return.
            bulk: Whether to return all items or the limit.
            available: Filter based on availability with operators.

        Returns:
        -------
            A list of Garage objects.
        """
        params: dict[str, Any] = {"limit": limit, "bulk": bulk}

        if available is not None:
            params["frei"] = str(available)

        locations = await self._request(
            "parkhaeuser/collections/verkehr_parkhaeuser/items",
            params=params,
        )

        # By default filter out garages without location coordinates.
        return [
            Garage.from_dict(item)
            for item in locations["features"]
            if item["geometry"] is not None
        ]

    async def close(self) -> None:
        """Close open client session."""
        if self.session and self._close_session:
            await self.session.close()

    async def __aenter__(self) -> UDPHamburg:
        """Async enter.

        Returns
        -------
            The Urban Data Platform object.
        """
        return self

    async def __aexit__(self, *_exc_info: str) -> None:
        """Async exit.

        Args:
        ----
            _exc_info: Exec type.
        """
        await self.close()
