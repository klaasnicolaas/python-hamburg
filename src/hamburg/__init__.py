"""Asynchronous Python client providing Urban Data information of Hamburg."""

from .exceptions import UDPHamburgConnectionError, UDPHamburgError
from .hamburg import UDPHamburg
from .models import DisabledParking, Garage, ParkAndRide

__all__ = [
    "UDPHamburg",
    "UDPHamburgConnectionError",
    "UDPHamburgError",
    "DisabledParking",
    "ParkAndRide",
    "Garage",
]
