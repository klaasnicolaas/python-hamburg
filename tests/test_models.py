"""Test the models."""
import aiohttp
import pytest
from aresponses import ResponsesMockServer

from hamburg import DisabledParking, ParkAndRide, UDPHamburg

from . import load_fixtures


@pytest.mark.asyncio
async def test_all_parking_spaces(aresponses: ResponsesMockServer) -> None:
    """Test all parking spaces function."""
    aresponses.add(
        "api.hamburg.de",
        "/datasets/v1/behindertenstellplaetze/collections/verkehr_behindertenparkpl/items",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/geo+json"},
            text=load_fixtures("disabled_parking.geojson"),
        ),
    )
    async with aiohttp.ClientSession() as session:
        client = UDPHamburg(session=session)
        spaces: list[DisabledParking] = await client.disabled_parkings()
        assert spaces is not None
        for item in spaces:
            assert isinstance(item, DisabledParking)
            assert item.spot_id is not None
            assert item.street is None or isinstance(item.street, str)
            assert item.limitation is None or isinstance(item.limitation, str)
            assert item.longitude is not None
            assert item.latitude is not None


@pytest.mark.asyncio
async def test_park_and_rides(aresponses: ResponsesMockServer) -> None:
    """Test park and ride spaces function."""
    aresponses.add(
        "api.hamburg.de",
        "/datasets/v1/p_und_r/collections/p_und_r/items",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/geo+json"},
            text=load_fixtures("park_and_ride.geojson"),
        ),
    )
    async with aiohttp.ClientSession() as session:
        client = UDPHamburg(session=session)
        spaces: list[ParkAndRide] = await client.park_and_ride()
        assert spaces is not None
        for item in spaces:
            assert item.spot_id is not None
            assert item.address is not None
            assert isinstance(item.tickets, dict)
            assert isinstance(item.longitude, float)
            assert isinstance(item.latitude, float)
