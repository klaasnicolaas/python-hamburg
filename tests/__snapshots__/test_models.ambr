# serializer version: 1
# name: test_all_parking_spaces
  list([
    DisabledParking(spot_id='7460', street='Steenwisch 105', limitation=None, number=1, longitude=9.946928367655032, latitude=53.58310925250252),
    DisabledParking(spot_id='7461', street='Langenfelder Damm 67', limitation=None, number=1, longitude=9.938008848629853, latitude=53.58105081830283),
    DisabledParking(spot_id='7462', street='Langenfelder Damm 80a', limitation=None, number=1, longitude=9.935971223239319, latitude=53.582530970209184),
    DisabledParking(spot_id='7463', street='Langenfelder Damm 91', limitation=None, number=1, longitude=9.93573271213673, latitude=53.582446224018476),
    DisabledParking(spot_id='7464', street='Stresemannstraße 318', limitation=None, number=1, longitude=9.9299854940711, latitude=53.563378917405764),
    DisabledParking(spot_id='7465', street='Bertrand-Russell-Straße 4', limitation=None, number=1, longitude=9.886326617358094, latitude=53.575552880265796),
    DisabledParking(spot_id='7466', street='Kiebitzstraße 44a', limitation=None, number=1, longitude=10.045029564340489, latitude=53.56500132834542),
    DisabledParking(spot_id='7467', street='Steintorplatz 2', limitation=None, number=1, longitude=10.008951895770426, latitude=53.55181415963955),
    DisabledParking(spot_id='7468', street='Steindamm 70', limitation=None, number=1, longitude=10.01523796793553, latitude=53.554727297628716),
    DisabledParking(spot_id='7469', street='Münzplatz 1', limitation=None, number=1, longitude=10.012553064840606, latitude=53.54923736617713),
  ])
# ---
# name: test_garages
  list([
    Garage(spot_id='1012', name='Suerhop', park_type='Parkplatz', disabled_parking_spaces=None, status='keine Auslastungsdaten', address=None, price=None, data_origin='HVV', free_space=None, capacity=20, availability_pct=None, longitude=9.854910321231317, latitude=53.315944994046816, updated_at=None),
    Garage(spot_id='1018', name='Dahlenburg', park_type='Straßenrand', disabled_parking_spaces=None, status='keine Auslastungsdaten', address=None, price=None, data_origin='HVV', free_space=None, capacity=6, availability_pct=None, longitude=10.70891715816324, latitude=53.16969724123711, updated_at=None),
    Garage(spot_id='1024', name='Ruschwedel', park_type='Straßenrand', disabled_parking_spaces=None, status='keine Auslastungsdaten', address=None, price=None, data_origin='HVV', free_space=None, capacity=6, availability_pct=None, longitude=9.559367386150887, latitude=53.446662778925074, updated_at=None),
    Garage(spot_id='1044', name='Echem', park_type='Parkplatz', disabled_parking_spaces=None, status='keine Auslastungsdaten', address=None, price=None, data_origin='HVV', free_space=None, capacity=18, availability_pct=None, longitude=10.53679025555922, latitude=53.33713455279623, updated_at=None),
    Garage(spot_id='1077', name='Tanneneck', park_type='Parkplatz', disabled_parking_spaces=None, status='keine Auslastungsdaten', address=None, price=None, data_origin='HVV', free_space=None, capacity=32, availability_pct=None, longitude=9.943234302538734, latitude=53.75551453553472, updated_at=None),
    Garage(spot_id='1094', name='Holstentherme', park_type='Parkplatz', disabled_parking_spaces=None, status='keine Auslastungsdaten', address=None, price=None, data_origin='HVV', free_space=None, capacity=None, availability_pct=None, longitude=9.945687641153029, latitude=53.83917387631545, updated_at=None),
    Garage(spot_id='1096', name='Kaltenkirchen Süd', park_type='Parkplatz', disabled_parking_spaces=None, status='keine Auslastungsdaten', address=None, price=None, data_origin='HVV', free_space=None, capacity=None, availability_pct=None, longitude=9.97323474361472, latitude=53.82324048644373, updated_at=None),
    Garage(spot_id='1097', name='dodenhof', park_type='Parkplatz', disabled_parking_spaces=None, status='keine Auslastungsdaten', address=None, price=None, data_origin='HVV', free_space=None, capacity=None, availability_pct=None, longitude=9.938193235695724, latitude=53.84447388984305, updated_at=None),
    Garage(spot_id='1106', name='Quickborner Straße', park_type='Parkplatz', disabled_parking_spaces=None, status='keine Auslastungsdaten', address=None, price=None, data_origin='HVV', free_space=None, capacity=22, availability_pct=None, longitude=9.989528817031388, latitude=53.7343090361397, updated_at=None),
    Garage(spot_id='1107', name='Richtweg', park_type='Straßenrand', disabled_parking_spaces=None, status='keine Auslastungsdaten', address=None, price=None, data_origin='HVV', free_space=None, capacity=20, availability_pct=None, longitude=9.990362195229995, latitude=53.695761978883716, updated_at=None),
  ])
# ---
# name: test_garages_live_data
  list([
    Garage(spot_id='10001', name='Alsterhaus', park_type='Parkhaus', disabled_parking_spaces=3, status='frei', address='Bei der Stadtwassermühle 16-20', price='Reguläres Parkentgelt 8:00 - 21:00 Uhr: 1. angef. Std. 4,00 EUR| je weitere Std. 4,00 EUR | Tageshöchstsatz: 28,00 EUR\nNachttarif 21:00 - 8:00 Uhr: 1. angef. Std. 1,50 EUR| je weitere Std. 1,50 EUR | Höchstsatz: 5,00 EUR', data_origin='BWVI_V', free_space=83, capacity=87, availability_pct=95.4, longitude=9.991765371268867, latitude=53.55266864038071, updated_at=datetime.datetime(2024, 2, 11, 3, 15, tzinfo=<DstTzInfo 'Europe/Berlin' CET+1:00:00 STD>)),
    Garage(spot_id='10002', name='Am Hauptbahnhof', park_type='Parkhaus', disabled_parking_spaces=2, status='frei', address='Borgesch 1', price='2,00 EUR/30min | Tageshöchstsatz 20,00 EUR| Nachttarif 1,50 EUR', data_origin='BWVI_V', free_space=341, capacity=547, availability_pct=62.3, longitude=10.00958322064467, latitude=53.55472284388092, updated_at=datetime.datetime(2024, 2, 11, 3, 15, tzinfo=<DstTzInfo 'Europe/Berlin' CET+1:00:00 STD>)),
    Garage(spot_id='10006', name='Parkhaus Stadthöfe (Bleichenhof)', park_type='Parkhaus', disabled_parking_spaces=2, status='frei', address='Große Bleichen 35', price='3,90 EUR/h|Tageshöchstsatz 32,00 EUR', data_origin='BWVI_V', free_space=216, capacity=590, availability_pct=36.6, longitude=9.986648065481441, latitude=53.55194593956319, updated_at=datetime.datetime(2024, 2, 11, 3, 15, tzinfo=<DstTzInfo 'Europe/Berlin' CET+1:00:00 STD>)),
    Garage(spot_id='10007', name='Radison Blu', park_type='Tiefgarage', disabled_parking_spaces=1, status='besetzt', address='Marseiller Straße 1', price='1. Std. 3,00 EUR|je weitere angef. Std. 3,00 EUR|Tagesentgelt 25,00 EUR\n', data_origin='BWVI_V', free_space=0, capacity=0, availability_pct=None, longitude=9.987726815763814, latitude=53.561135227527565, updated_at=datetime.datetime(2024, 2, 11, 3, 15, tzinfo=<DstTzInfo 'Europe/Berlin' CET+1:00:00 STD>)),
    Garage(spot_id='10010', name='Deichtorhallen', park_type='Parkplatz', disabled_parking_spaces=2, status='frei', address='Deichtorstraße 2', price='1 Stunde 2,00 EUR|Tageshöchstsatz 20,00 EUR|Nachttarif 22 - 6 Uhr 1,00 EUR', data_origin='BWVI_V', free_space=35, capacity=37, availability_pct=94.6, longitude=10.00496859112039, latitude=53.54689636607786, updated_at=datetime.datetime(2024, 2, 11, 3, 15, tzinfo=<DstTzInfo 'Europe/Berlin' CET+1:00:00 STD>)),
    Garage(spot_id='10011', name='Deutsch Japanisches Zentrum', park_type='Tiefgarage', disabled_parking_spaces=12, status='frei', address='Stadthausbrücke 1', price='Reguläres Parkentgelt: erste 30 Min. 1,50 EUR |zweite 30 Min. 1,50 EUR |Je weitere Std. 3,00 EUR |Tag (24 Std.) 22,00 EUR\nSondertarif (P-Card): erste 30 Min. 0,50 EUR |zweite 30 Min. 1,00 EUR |Je weitere Std. 1,50 EUR |Tag (24 Std.) 9,00 EUR', data_origin='BWVI_V', free_space=240, capacity=291, availability_pct=82.5, longitude=9.985204026496357, latitude=53.55121608632646, updated_at=datetime.datetime(2024, 2, 11, 3, 15, tzinfo=<DstTzInfo 'Europe/Berlin' CET+1:00:00 STD>)),
    Garage(spot_id='10012', name='Elbphilharmonie', park_type='Parkhaus', disabled_parking_spaces=23, status='frei', address='Platz der Deutschen Einheit 1', price='Parkbereich A Kurzparker (Ebenen OG1 – OG5): 1. Stunde 5,00 EUR|jede weitere Std. 5,00 EUR|Tageshöchstsatz 50,00 EUR & Parkbereich B (UG1 = „Hotel-Bereich“): 1. Stunde 6,00 EUR|jede weitere Std. 6,00 EUR|Tageshöchstsatz 50,00 EUR & Parkbereich C (EG = Umfahrung): 5 Minuten freie |Durchfahrt | je weitere 5 Minuten 1,00 EUR | Tageshöchstsatz 90,00 EUR', data_origin='BWVI_V', free_space=243, capacity=277, availability_pct=87.7, longitude=9.984820179727356, latitude=53.54130630618613, updated_at=datetime.datetime(2024, 2, 11, 3, 15, tzinfo=<DstTzInfo 'Europe/Berlin' CET+1:00:00 STD>)),
    Garage(spot_id='10013', name='Europa-Passage', park_type='Parkhaus', disabled_parking_spaces=31, status='frei', address='Hermannstraße 09 - 11', price='Reguläres Parkentgelt: 1. Std 2,50 EUR |jede weitere Std. 2,50 EUR |Tag (24 Std.) 16,00 EUR\nSondertarif (P-Card): 1. Std 2,00 EUR |jede weitere Std. 2,00 EUR |Tag (24 Std.) 12,00 EUR', data_origin='BWVI_V', free_space=670, capacity=700, availability_pct=95.7, longitude=9.99667711811397, latitude=53.551947553640574, updated_at=datetime.datetime(2024, 2, 11, 3, 15, tzinfo=<DstTzInfo 'Europe/Berlin' CET+1:00:00 STD>)),
    Garage(spot_id='10015', name='City-Parkhaus', park_type='Parkhaus', disabled_parking_spaces=None, status='frei', address='Gertrudenstraße 2', price='3,30 EUR/h|Tageshöchstsatz 20,00 EUR', data_origin='BWVI_V', free_space=378, capacity=1110, availability_pct=34.1, longitude=10.0007919727711, latitude=53.552931554513535, updated_at=datetime.datetime(2024, 2, 11, 3, 15, tzinfo=<DstTzInfo 'Europe/Berlin' CET+1:00:00 STD>)),
    Garage(spot_id='10016', name='Große Reichenstraße', park_type='Parkhaus', disabled_parking_spaces=3, status='frei', address='Große Reichenstraße 14', price='2,00 EUR/30min |Tageshöchstsatz 20,00 EUR|Nachttarif 1,50 EUR\n', data_origin='BWVI_V', free_space=441, capacity=965, availability_pct=45.7, longitude=9.996296605176623, latitude=53.54832020044909, updated_at=datetime.datetime(2024, 2, 11, 3, 15, tzinfo=<DstTzInfo 'Europe/Berlin' CET+1:00:00 STD>)),
  ])
# ---
# name: test_park_and_rides
  list([
    ParkAndRide(spot_id='3656903', name='Volksdorf', park_type='Parkhaus', address='Farmsener Landstraße 202, 22359 Hamburg', construction_year=None, public_transport_line='U1', disabled_parking_spaces=3, tickets={'day': '0', 'month': '0', 'year': '0'}, url=None, free_space=349, capacity=350, availability_pct=99.7, longitude=10.161686000000119, latitude=53.64966699999995, updated_at=datetime.datetime(2024, 2, 11, 0, 40, 56, tzinfo=<DstTzInfo 'Europe/Berlin' CET+1:00:00 STD>)),
    ParkAndRide(spot_id='3656904', name='Meiendorfer Weg', park_type='Parkhaus', address='Meiendorfer Weg 127, 22359 Hamburg', construction_year=None, public_transport_line='U1', disabled_parking_spaces=3, tickets={'day': '0', 'month': '0', 'year': '0'}, url=None, free_space=226, capacity=254, availability_pct=89.0, longitude=10.155718000000116, latitude=53.63889299999997, updated_at=datetime.datetime(2024, 2, 11, 2, 7, 48, tzinfo=<DstTzInfo 'Europe/Berlin' CET+1:00:00 STD>)),
    ParkAndRide(spot_id='3656905', name='Trabrennbahn', park_type='Parkplatz', address='Traberweg, 22159 Hamburg', construction_year='1966', public_transport_line='U1', disabled_parking_spaces=0, tickets={'day': '2', 'month': '20', 'year': '100'}, url='www.pr.hamburg', free_space=79, capacity=90, availability_pct=87.8, longitude=10.102240000000092, latitude=53.59862799999995, updated_at=datetime.datetime(2024, 2, 11, 0, 58, 6, tzinfo=<DstTzInfo 'Europe/Berlin' CET+1:00:00 STD>)),
    ParkAndRide(spot_id='3656906', name='Poppenbüttel', park_type='Parkplatz', address='Tennigkeitweg 1b, 22391 Hamburg', construction_year=None, public_transport_line='S1, S11', disabled_parking_spaces=3, tickets={'day': '0', 'month': '0', 'year': '0'}, url=None, free_space=75, capacity=90, availability_pct=83.3, longitude=10.092609000000087, latitude=53.65071799999995, updated_at=datetime.datetime(2024, 2, 11, 2, 18, 36, tzinfo=<DstTzInfo 'Europe/Berlin' CET+1:00:00 STD>)),
    ParkAndRide(spot_id='3656907', name='Poppenbüttel', park_type='Parkhaus', address='Stormarnplatz 3, 22393 Hamburg', construction_year=None, public_transport_line='S1, S11', disabled_parking_spaces=3, tickets={'day': '0', 'month': '0', 'year': '0'}, url=None, free_space=309, capacity=365, availability_pct=84.7, longitude=10.094348000000089, latitude=53.65207799999995, updated_at=datetime.datetime(2024, 2, 11, 3, 5, 59, tzinfo=<DstTzInfo 'Europe/Berlin' CET+1:00:00 STD>)),
    ParkAndRide(spot_id='3656908', name='Berne', park_type='Parkpalette', address='Berner Heerweg, 22159 Hamburg', construction_year=None, public_transport_line='U1', disabled_parking_spaces=4, tickets={'day': '0', 'month': '0', 'year': '0'}, url=None, free_space=153, capacity=187, availability_pct=81.8, longitude=10.139647000000108, latitude=53.62727899999995, updated_at=datetime.datetime(2024, 2, 11, 2, 59, 38, tzinfo=<DstTzInfo 'Europe/Berlin' CET+1:00:00 STD>)),
    ParkAndRide(spot_id='3656909', name='Steinfurther Allee', park_type='Parkhaus', address='Möllner Landstraße 217, 22117 Hamburg', construction_year=None, public_transport_line='U2', disabled_parking_spaces=3, tickets={'day': '0', 'month': '0', 'year': '0'}, url=None, free_space=337, capacity=345, availability_pct=97.7, longitude=10.13845800000011, latitude=53.541055999999955, updated_at=datetime.datetime(2024, 2, 11, 2, 10, 44, tzinfo=<DstTzInfo 'Europe/Berlin' CET+1:00:00 STD>)),
    ParkAndRide(spot_id='3656910', name='Rahlstedt', park_type='Parkhaus', address='Doberaner Weg 16, 22143 Hamburg', construction_year='2003', public_transport_line='RB81', disabled_parking_spaces=3, tickets={'day': '2', 'month': '20', 'year': '100'}, url='www.pr.hamburg', free_space=122, capacity=316, availability_pct=38.6, longitude=10.151751000000116, latitude=53.60431399999996, updated_at=datetime.datetime(2024, 1, 17, 12, 11, 22, tzinfo=<DstTzInfo 'Europe/Berlin' CET+1:00:00 STD>)),
    ParkAndRide(spot_id='3656911', name='Bergedorf', park_type='Parkhaus', address='Bergedorfer Strasse 85a, 21033 Hamburg', construction_year=None, public_transport_line='S2, S21', disabled_parking_spaces=4, tickets={'day': '0', 'month': '0', 'year': '0'}, url=None, free_space=359, capacity=403, availability_pct=89.1, longitude=10.203916000000143, latitude=53.48942299999997, updated_at=datetime.datetime(2024, 2, 11, 3, 10, 11, tzinfo=<DstTzInfo 'Europe/Berlin' CET+1:00:00 STD>)),
    ParkAndRide(spot_id='3656912', name='Nettelnburg Nord', park_type='Parkplatz', address='Friedrich Frank Bogen 109, 21033 Hamburg', construction_year=None, public_transport_line='S2, S21', disabled_parking_spaces=0, tickets={'day': '0', 'month': '0', 'year': '0'}, url=None, free_space=213, capacity=236, availability_pct=90.3, longitude=10.182725000000133, latitude=53.488834999999966, updated_at=datetime.datetime(2024, 2, 11, 1, 16, 26, tzinfo=<DstTzInfo 'Europe/Berlin' CET+1:00:00 STD>)),
  ])
# ---
